docs/guide/rules.md

```diff
@@ -0,0 +1,316 @@
+# Rules
+
+## `Required` - requiring values
+
+Use `Yiisoft\Validator\Rule\Required` rule to make sure value is provided. What values are considered empty can be
+customized via `$emptyCallback` option. Normalization is not performed here, so only a callable or special class is
+needed. For more details see "Skipping empty values" section.
+
+```php
+use Yiisoft\Validator\Rule\Required;
+use Yiisoft\Validator\EmptyCriteria\WhenNull;
+
+$rules = [new Required(emptyCallback: new WhenNull())];
+```
+
+## `Nested` - nested and related data
+
+### Basic usage
+
+In many cases there is a need to validate related data in addition to current entity / model. There is a `Nested` rule
+for this purpose:
+
+```php
+use Yiisoft\Validator\ValidatorInterface;
+use Yiisoft\Validator\Rule\HasLength;
+use Yiisoft\Validator\Rule\Nested;
+use Yiisoft\Validator\Rule\Number;
+use Yiisoft\Validator\Rule\Required;
+
+// Usually obtained from container
+$validator = $container->get(ValidatorInterface::class);
+
+$data = ['author' => ['name' => 'Alexey', 'age' => '31']];
+$rule = new Nested([
+    'title' => [new Required()],
+    'author' => new Nested([
+        'name' => [new HasLength(min: 3)],
+        'age' => [new Number(min: 18)],
+    )];
+]);
+$errors = $validator->validate($data, [$rule])->getErrorMessagesIndexedByPath();
+```
+
+### Other configuration options
+
+A dot notation can be used as an alternative way of configuration. In this case the example above will be presented as
+following:
+
+```php
+use Yiisoft\Validator\Rule\HasLength;
+use Yiisoft\Validator\Rule\Nested;
+use Yiisoft\Validator\Rule\Number;
+use Yiisoft\Validator\Rule\Required;
+
+$rule = new Nested([
+    'title' => [new Required()],
+    'author.name' => [new HasLength(min: 3)],
+    'author.age' => [new Number(min: 18)],
+)];
+```
+
+It's also possible to combine both of these approaches:
+
+```php
+use Yiisoft\Validator\Rule\HasLength;
+use Yiisoft\Validator\Rule\Nested;
+use Yiisoft\Validator\Rule\Number;
+use Yiisoft\Validator\Rule\Required;
+
+$data = ['author' => ['name' => 'Alexey', 'age' => '31']];
+$rule = new Nested([
+    'author' => new Nested([
+        'name' => [new HasLength(min: 3)],
+        'age' => [new Number(min: 18)],
+    )];
+]);
+```
+
+### Advanced usage
+
+A more complex real-life example is a chart that is made of points. This data is represented as arrays. `Nested` can be
```

Comment:
Would be cool to add a line chart image here but that could be done later on the "enhancements" stage.

Reply:
Added to TODO list (optional).

Reply:
Decided to add it in this PR.

Reply:
Done, please check.

Reply:
👍